### DISCUSSION TITLE
Dark Mode Not Working In Deployed Env

### DISCUSSION THREAD

**Discussion by jamesmart77** [darkreader/darkreader#8324] [Q&A]
Hey all, I'm pumped about this library! Something super weird happens in my application whee the darkmode works perfectly locally, but when I deploy it most of the page does not get updated. Any thoughts about what that could be?

Working Pic (locally)
![image](https://user-images.githubusercontent.com/29242164/156403177-57ff5eb6-069d-4894-9ee2-7293a98452c9.png)

Not working (deployed)
![image](https://user-images.githubusercontent.com/29242164/156402878-5416cb15-1c33-4330-b037-359007e0e560.png)

**Comment by Gusted** [comment]
Does the dev tools generate any errors or warnings?

**Reply by jamesmart77** [reply]
Unfortunately, no :\

**Reply by Gusted** [reply]
Is the site publicly accessible?

**Reply by jamesmart77** [reply]
Yea i wish i could share it, but it's behind a company firewall :( Let me know if there is anything I can send you for helpful info

**Reply by Gusted** [reply]
Okay so:

- Is `DarkReader` a object in the global scope?(e.g. does `DarkReader` output anything if you enter it in the console and press enter).
- What does `DarkReader.isEnabled()` output?
- What kind of code is trying to enable Dark Reader(if you `console.log("I'm here!")` before the `DarkReader.enable(...)` does it show that in console?)

**Reply by jamesmart77** [reply]
Doesn't look like `DarkReader` is referenced. I can see though in the `html` tag that the attributes are being set
![image](https://user-images.githubusercontent.com/29242164/156416182-727dfe2d-876c-4173-a142-8f5ff8d288a1.png)

**Reply by Gusted** [reply]
Do you have by any chance still the dark reader extension enabled? In order to test the API correctly it's best to disable it or whitelist the site.

**Reply by jamesmart77** [reply]
Idk if this is helpful, but in a separate implementation, it looks like SOME of the page elements are being updated when darkMode is enabled. You can see below the "Web Accessibility" header, the light grey text in the footer is being modified. Weird!
![image](https://user-images.githubusercontent.com/29242164/156416594-ea479c16-e0e4-4313-82ec-b980d2346ec4.png)

**Reply by jamesmart77** [reply]
I never actually installed the chrome extension and don't have any other extensions for dark mode

**Reply by Gusted** [reply]
Hmm, that's weird how is Dark reader being loaded? By default it will nest itself onto the global scope.

**Reply by jamesmart77** [reply]
Weird indeed, esp since all is good locally but not when deployed.

I have the `useDarkReader` hook being called on render here
![image](https://user-images.githubusercontent.com/29242164/156434446-bc212ac4-9f4a-410f-9792-d4520bf01d31.png)

And here's the `userDarkReader` hook
![image](https://user-images.githubusercontent.com/29242164/156434577-a94f032a-ee7a-454d-b856-7581b61aee70.png)

And because `darkreader` is a dependency in this project (which is a npm package), i have it declared in the webpack externals
![image](https://user-images.githubusercontent.com/29242164/156434905-32209572-db4e-4ab7-bcf2-0f5ca89dd1d5.png)

**Reply by Gusted** [reply]
I'm not sure if I'm much of help here, it sounds like a problem with webpack or with the npm package. If there is no error from dark reader then it should be working.

**Reply by Gusted** [reply]
Not sure if your build configuration differs locally vs production, otherwise it could be that there are CORS/CSP errors going on in dark reader. Check if any network request fails.

**Reply by jamesmart77** [reply]
@Gusted i appreciate the help. 

I was poking around trying to recreate this locally by following your steps [referenced above](https://github.com/darkreader/darkreader/discussions/8324#discussioncomment-2282099) but cannot get the console to output anything...even though the darkmode works 🤔  Any ideas?

Here is a screenshot of the hook setting up the darkreader and logging out the object. As you can see in the logs it's `undefined` and then when i try to reference the global `DarkReader` object in the next console statement that is not defined either...but darkmode works locally??
![image](https://user-images.githubusercontent.com/29242164/156783789-3aaff208-771c-45e9-9ea8-733562d07b4c.png)

Here's a screenshot of darkmode working locally
![image](https://user-images.githubusercontent.com/29242164/156784406-2e9d4dd1-6863-4a71-9afd-1c9043d7b7ef.png)

**Reply by Gusted** [reply]
Seems like using the ESM requires a bit different... Import the IsEnabled function from dark reader and try to log the output of that. Did you check if you have network requests failing when on production?

**Reply by jamesmart77** [reply]
Here you can see in the console logs that `DarkReader.isEnabled()` is printing from within the custom hook i have. And you can see that part of the page is being set to darkmode (ie. the footer, the accessibility menu popup).
![image](https://user-images.githubusercontent.com/29242164/156799516-2f770131-51df-46b7-bda2-41e88c852fff.png)

I can also see that the custom `style` tags are being added to the document.head as expected when darkmode is toggled on.
![image](https://user-images.githubusercontent.com/29242164/156799908-dad3e07a-ac13-4621-adb4-88c6393da310.png)

Are there any specific setup instructions you can think of that would be causing a "partial" darkmode scenario?

**Reply by jamesmart77** [reply]
Also, not seeing any network issues -- looks like everything is 200. Which would make sense given it's a partial darkmode situation.

**Reply by jamesmart77** [reply]
Well this is interesting, the first `body { background-color: white; }` is the style set within the application using the `darkreader` and way down at the bottom of the shot you can see the `html, body { background-color: #222425; }` set by the darkreader in the `user-agent` `style` tag (see second screenshot). Since styling order matters, any ideas how to have the app styles fall before the darkreader styles in this react app?

![image](https://user-images.githubusercontent.com/29242164/156810264-2104858e-809d-4334-b46d-6ca5fc4eb353.png)

Second screenshot
![image](https://user-images.githubusercontent.com/29242164/156810783-81a3da08-3594-4c14-8458-bab0e8257b65.png)

**Reply by jamesmart77** [reply]
Sorry for all the messages, but i think i figured out the root cause, now just need to find a solution.

Looks like the order of the `<style>` tags is messing it all up. In the local env, there is no `<script src="/static/js/main.35da339b.js" />` which is why the darkreader works as expected. But during the build, this script tag gets placed **after** the darkreader scripts which explains why only **some** of the darkmode styles are taking effect. 

![image](https://user-images.githubusercontent.com/29242164/156812102-d06cc050-2760-44f7-8c96-a690a5809feb.png)

I don't suppose you know of an easy fix? Looks like those `<style>` tags are appended first on the body rather than at the end...do you know why they're not appended at the end so that all darkreader styles are the last pieces to get appended?
![image](https://user-images.githubusercontent.com/29242164/156812633-8dc6543f-dfd5-43d8-b598-3515e620dbb8.png)

**Reply by Gusted** [reply]
> I don't suppose you know of an easy fix? Looks like those <style> tags are appended first on the body rather than at the end...do you know why they're not appended at the end so that all darkreader styles are the last pieces to get appended?

They are "static" CSS, they are the basis for dark reader, but a `--sync` should without any problems be able to override such basic fix.


I don't think I'm understanding the role of `<script>` here and why it could cause a issue... Are you able to move such script to e.g. `<body>`? To confirm its' that specific cause.

**Reply by jamesmart77** [reply]
Apologies, not `<script>` but the `<link href="static/css/main....css" rel="stylesheet" />`. That is the css generated from the react-build process.

**Reply by Gusted** [reply]
Well yes, that doesn't get modified? Well then I cannot really help any further as that could be for a thousand of reasons, mainly because there is no `--sync` style being appended(which is a very simple step once a new style has been found).

**Comment by jamesmart77** [comment]
Ok, thanks for all the help. Last question, is the `--sync` already part of
the process or something I need to add?

**Reply by Gusted** [reply]
> Ok, thanks for all the help. Last question, is the `--sync` already part of the process or something I need to add?

It's added once dark reader sees a new style(as it should be doing in this case, but somehow isn't).
